data-prepper-plugins/s3-source/src/main/java/org/opensearch/dataprepper/plugins/source/ScanOptions.java

```diff
@@ -99,36 +82,68 @@ public Builder setRange(Duration range) {
             return this;
         }
 
-        public Builder setBucket(String bucket) {
-            this.bucket = bucket;
+        public Builder setEndDateTime(LocalDateTime endDateTime) {
+            this.endDateTime = endDateTime;
             return this;
         }
 
-        public Builder setEndDateTime(LocalDateTime endDateTime) {
-            this.endDateTime = endDateTime;
+        public Builder setBucketOption(S3ScanBucketOption bucketOption) {
+            this.bucketOption = bucketOption;
             return this;
         }
-        public ScanOptions build() {
-            if(Objects.nonNull(startDateTime) && Objects.nonNull(range) && Objects.nonNull(endDateTime))
-                scanRangeDateValidationError();
 
-            if(Objects.nonNull(startDateTime) && Objects.nonNull(endDateTime)){
+        public ScanOptions build() {
+            // Check for bucket-specific time range first
+            LocalDateTime bucketStartDateTime = bucketOption.getStartTime();
+            LocalDateTime bucketEndDateTime = bucketOption.getEndTime();
+            Duration bucketRange = bucketOption.getRange();
+            int nonNullCount = (Objects.isNull(bucketStartDateTime) ? 0 : 1) + (Objects.isNull(bucketRange) ? 0 : 1)
+                    + (Objects.isNull(bucketEndDateTime) ? 0 : 1);
+            if (nonNullCount == 1 || nonNullCount == 3) {
+                scanRangeDateValidationError(bucketOption.getName());
```

Comment:
Won't this throw an exception if the bucket doesn't have these values configured before we have the chance to check the global settings? What do we validate on the config immediately?

The code is also a bit confusing. Can we change the flow to this


```
LocalDateTime bucketStartDateTime = Objects.isNull(bucketOption.getStartDateTime()) ? globalStartTime : bucketOption.getStartDateTime(); 
LocalDateTime bucketEndDateTime = Objects.isNull(bucketOption.getEndDateTime()) ? globalEndTime : bucketOption.getStartDateTime(); 
Duration bucketRange = Objects.nonNull(bucketOption.getRange()) ? globalRange : bucketOption.getRange();

validateBucketOptions(bucketStartDateTime, bucketEndDateTime, bucketRange);

// construct based on range, start, and end combo
```

Reply:
This has the benefit of allowing things like

```
start_time: "start"
bucket:
   - name: "bucket-one:
      end_time: "end"
   - name: "bucket-two"
      end_time: "different-end"
```

Reply:
Interesting! I didn't think about combining global and bucket settings together. 

Does it work as intended in this case though?
```
start_time: "start"
end_time: "end"
bucket:
   - name: bucket-one
      start_time: "bucket_start"
      range: "bucket_range"
   - name: bucket-two
```
This looks like a valid setting to me - bucket-one uses its own setting while bucket-two uses global setting. 
But we would feed three non-nulls into `validateBucketOptions(bucketStartDateTime, bucketEndDateTime, bucketRange)` for the first bucket and get a validation error I assume?

Reply:
That's a good point. Yes what you have above should be valid. I think we can modify the code a bit for this.

```
LocalDateTime bucketStartDateTime = Objects.isNull(bucketOption.getStartDateTime()) ? globalStartTime : bucketOption.getStartDateTime(); 
LocalDateTime bucketEndDateTime = Objects.isNull(bucketOption.getEndDateTime()) ? globalEndTime : bucketOption.getStartDateTime(); 
Duration bucketRange = Objects.nonNull(bucketOption.getRange()) ? globalRange : bucketOption.getRange();

validateHasAtLeastTwoNonNull(start, end, range);

if (atLeastOneIsConfiguredOnBucket()) {
      if (hasTwoConfiguredOnBucket) {
            setTheThirdToNull();
       } else { // has only one on bucket
           setOneOfTheGlobalOptionsToNull();
      }
}
```

Reply:
Thanks! That works for most cases. I reworked it a bit in latest commit and added more test cases.